commit 1df1666745106f32b2b3e94633774b1c602a3ebe (refs/changes/77/5077/2)
Author: Josh Wilsdon <josh@wilsdon.ca>
Date:   2018-11-21T12:43:33-08:00 (11 months ago)
    
    TRITON-972 cn-agent-update service should not register itself and should not send heartbeats
    Reviewed by: Pedro Palazón Candel <pedro@joyent.com>
    Approved by: Pedro Palazón Candel <pedro@joyent.com>

diff --git a/lib/app.js b/lib/app.js
index 38c1c06..4f2a128 100644
--- a/lib/app.js
+++ b/lib/app.js
@@ -84,6 +84,14 @@ function App(options) {
         ' server/' + options.uuid;
     this.uuid = options.uuid;
 
+    this.isUpdateHelper = false;
+    if (process.env.SMF_FMRI === 'svc:/smartdc/agent/cn-agent-update:default') {
+        // If we're the cn-agent-update service, we won't register with CNAPI
+        // since the cn-agent service will do that and the cn-agent-update
+        // service is only ever used to by cn-agent itself.
+        this.isUpdateHelper = true;
+    }
+
     // ensure we have all the config required to connect to and use CNAPI
     assert.object(this.config, 'this.config');
     assert.optionalObject(this.config.cnapi, 'this.config.cnapi');
@@ -467,6 +475,20 @@ App.prototype.start = function () {
         queueDefns[i].onhttpmsg = createHttpTaskDispatchFn(agent, taskspath);
     }
 
+    function completeStartup() {
+        agent.useQueues(queueDefns);
+        self.log.info('starting cn-agent for %s', self.uuid);
+        agent.start();
+    }
+
+    if (self.isUpdateHelper) {
+        // As described above, we don't want to start the heartbeater or clean
+        // up locks that are not ours here. So we just start listening for
+        // tasks.
+        completeStartup();
+        return;
+    }
+
     vasync.pipeline({
         funcs: [
             function _createCnapiConnection(_, cb) {
@@ -492,9 +514,7 @@ App.prototype.start = function () {
             throw err;
         }
 
-        agent.useQueues(queueDefns);
-        self.log.info('starting cn-agent for %s', self.uuid);
-        agent.start();
+        completeStartup();
     });
 };
 
diff --git a/package.json b/package.json
index 7d884da..9bf9ac9 100644
--- a/package.json
+++ b/package.json
@@ -1,7 +1,7 @@
 {
   "name": "cn-agent",
   "description": "Triton Compute Node Agent",
-  "version": "2.3.2",
+  "version": "2.4.0",
   "author": "Joyent (joyent.com)",
   "private": true,
   "dependencies": {
